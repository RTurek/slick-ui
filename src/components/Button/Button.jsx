import React from 'react';
import PropTypes from 'prop-types';
import { useTheme } from '@material-ui/core/styles';
import styled, { css } from 'styled-components';
import classNames from 'classnames';
import { rgba } from 'polished';
import MaterialButton from '@material-ui/core/Button';
import getColorFromTheme from '../../styles/getColorFromTheme';
import { BUTTON_SIZES } from '../../styles/sizes';

/*
TODO: Refactor color choices for button to be an array of const values...
TODO: ...that can be exported to other files for use in PropTypes, Storybook, and tests
*/

/** checks to see if the color is default of inherit - set it to */
function defaultColors(color) {
  return ['default', 'inherit'].includes(color) ? 'secondary' : color;
}

function validVariant(variant) {
  if (variant === 'light') {
    return 'outlined';
  }
  return variant;
}

function sizeAndFontCss(size, iconPadding) {
  switch (size) {
    case BUTTON_SIZES.small:
      return `
        font-size: .875rem;
        padding: ${iconPadding ? '4px' : '7px 13px'};
        min-width: ${iconPadding ? '10px' : '120px'};
      `;
    case BUTTON_SIZES.large:
      return `
        font-size: 1.125rem;
        padding: ${iconPadding ? '12px' : '13px 18px'};
        min-width: ${iconPadding ? '10px' : '120px'};
      `;
    default: // Medium
      return `
        font-size: 1rem;
        padding: ${iconPadding ? '8px' : '10px 16px'};
        min-width: ${iconPadding ? '10px' : '120px'};
      `;
  }
}

function basicCss(color, size, iconPadding) {
  return `
    transition: box-shadow 0.3s ease-in-out, transform 0.3s ease-in-out;
    color: ${color};
    font-family: Lato;
    text-transform: capitalize;
    font-style: normal;
    font-weight: 800;
    line-height: 18px;
    letter-spacing: 0.02em;
    ${sizeAndFontCss(size, iconPadding)}
  `;
}

function focusCss(square, colorVariant) {
  return `
    &:focus {
      border-radius: ${square ? 8 : 999}px;
      box-shadow: 0 0 0 4px ${rgba(colorVariant, 0.2)};
    }
  `;
}

function hoverAndPopCss(square, colorVariant, makeItPop) {
  return `
    &:hover {
      transform: ${makeItPop ? 'scale(1.1)' : 'scale(1)'};
      border-radius: ${square ? 8 : 999}px;
      box-shadow: 0 0 0 4px ${rgba(colorVariant, 0.2)};
    }
  `;
}

const StyledButton = styled(({
  square, variant, color, textColor, iconPadding, makeItPop, ...rest
}) =>
  <MaterialButton variant={validVariant(variant)} {...rest}/>)`
&&& {
  .MuiButton-startIcon {
    margin-left: 0px;
  }
  .MuiButton-endIcon {
    margin-right: 0px;
  }
  border-radius: ${({ square }) => (square ? 8 : 999)}px;
  min-width: 120px;
  max-width: 288px;
  color: #FFFFFF;
  ${({ variant, color, makeItPop, disabled, textColor, square, size, iconPadding }) => {
    const theme = useTheme();
    const colorVariant = getColorFromTheme(theme, defaultColors(color)) || 'white';
    const textColorVariant = getColorFromTheme(theme, defaultColors(textColor)) || 'white';
    switch (variant) {
      case 'contained':
        return css`
          background-color: ${colorVariant};
          box-shadow: unset;
          ${basicCss(textColorVariant, size, iconPadding)}
          ${focusCss(square, colorVariant)}
          ${hoverAndPopCss(square, colorVariant, makeItPop)}
          ${disabled && `
            background-color: ${theme.palette.backgroundColor.main} !important;
            color: ${theme.palette.lightGrey.main} !important;
          `}
        `;
      case 'outlined':
        return css`
          background-color: inherit;
          border: 2px solid ${rgba(colorVariant, 0.4)};
          ${basicCss(colorVariant, size, iconPadding)}
          ${focusCss(square, colorVariant)}
          ${hoverAndPopCss(square, colorVariant, makeItPop)}
          ${disabled && `
            border: 2px solid ${rgba(theme.palette.backgroundColor.main, 0.7)};
            color: ${theme.palette.lightGrey.main} !important;
            box-shadow: unset;
          `}
        `;
      case 'light':
        return css`
          border: none;
          background-color: ${rgba(colorVariant, 0.1)};
          ${basicCss(colorVariant, size, iconPadding)}
          ${focusCss(square, colorVariant)}
          ${hoverAndPopCss(square, colorVariant, makeItPop)}
          ${disabled && `
            background-color: ${rgba(theme.palette.backgroundColor.main, 0.4)};
            color: ${theme.palette.lightGrey.main} !important;
            box-shadow: unset;
          `}
        `;
      case 'text':
        return css`
          background-color: inherit;
          ${basicCss(colorVariant, size, iconPadding)}
          ${focusCss(square, colorVariant)}
          ${hoverAndPopCss(square, colorVariant, makeItPop)}
          ${disabled && `
            color: ${theme.palette.lightGrey.main} !important;
            box-shadow: unset;
          `}
        `;
      default:
        return '';
    }
  }}
}
`;

function Button(props) {
  const {
    className, inputProps, ...otherProps
  } = props;
  let spreadIP;
  if (inputProps || otherProps['data-testid']) {
    spreadIP = {
      ...otherProps['data-testid'] && { 'data-testid': otherProps['data-testid'] },
      ...inputProps
    };
  }
  return (
    <StyledButton
      className={classNames(
        'react-css',
        className
      )}
      {...spreadIP}
      {...otherProps}
    />
  );
}

Button.propTypes = {
  /** variant */
  variant: PropTypes.oneOf(['contained', 'outlined', 'light', 'text']),
  /** color */
  color: PropTypes.string,
  /** size, one of three options */
  size: PropTypes.oneOf([BUTTON_SIZES.small, BUTTON_SIZES.medium, BUTTON_SIZES.large]),
  /** textColor */
  textColor: PropTypes.string,
  /** disabled */
  disabled: PropTypes.bool,
  /** add class */
  className: PropTypes.string,
  /** inputProps */
  inputProps: PropTypes.shape({
  }),
  /** sets border radius to 8px */
  square: PropTypes.bool,
  /** adds hover elevation to contained variant */
  makeItPop: PropTypes.bool,
  /** sets padding to 6px so icon-only buttons look better */
  iconPadding: PropTypes.bool
};

Button.defaultProps = {
  variant: 'contained',
  color: 'primary',
  size: BUTTON_SIZES.medium,
  className: '',
  disabled: false,
  textColor: undefined,
  inputProps: undefined,
  square: false,
  makeItPop: false,
  iconPadding: false
};

export default Button;
