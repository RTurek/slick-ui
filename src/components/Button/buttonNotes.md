#### Button
---

This is a wrapper component for Material Button.

#### Notes
---
Some fun custom styling has been added, such as "makeItPop" which applies hoverAndPopCss to make the button stand out a bit more on hover/click.

#### Sources
---
- https://mui.com/components/buttons/
