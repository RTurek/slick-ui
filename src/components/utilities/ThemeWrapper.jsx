import React from 'react';
import PropTypes from 'prop-types';
import { createTheme } from '@material-ui/core/styles';
import { ThemeProvider } from '@material-ui/styles';
import { COLORS } from '@styles/colors';

class ThemeWrapper extends React.Component {
  constructor(props) {
    super(props);

    const primaryThemeColor = COLORS.primary
    const secondaryThemeColor = COLORS.secondary
    const textPrimaryThemeColor = COLORS.textPrimary;
    const textSecondaryThemeColor = COLORS.textSecondary;
    const backgroundThemeColor = COLORS.background;
    const alertThemeColor = COLORS.alert;
    const urgentThemeColor = COLORS.urgent;
    const successThemeColor = COLORS.success;
    const warningThemeColor = COLORS.warning;
    const crGreenThemeColor = COLORS.crGreen;
    const rpOrangeThemeColor = COLORS.rpOrange;
    const darkGreyThemeColor = COLORS.darkGrey;
    const mediumGreyThemeColor = COLORS.mediumGrey;
    const lightGreyThemeColor = COLORS.lightGrey;
    const modernBlackThemeColor = COLORS.modernBlack;

    this.muiTheme = createTheme({
      typography: {
        // fontFamily: primaryFontFamily,
        useNextVariants: true,
        h1: {
          letterSpacing: '-1.5px'
        },
        h2: {
          letterSpacing: '-0.5px'
        },
        h3: {
          lineHeight: 1,
          letterSpacing: '0px',
          fontWeight: '300'
        },
        h4: {
          fontSize: '1.875rem',
          letterSpacing: '0.25px',
          lineHeight: 1.0667
        },
        h5: {
          letterSpacing: '0px',
          lineHeight: 1.0833
        },
        h6: {
          letterSpacing: '0.15px',
          lineHeight: 1.1
        },
        subtitle1: {
          letterSpacing: '0.15px',
          lineHeight: 1.125
        },
        subtitle2: {
          letterSpacing: '0.1px',
          lineHeight: 1.1428
        },
        body1: {
          letterSpacing: '0.25px',
          lineHeight: 1.25
        },
        body2: {
          letterSpacing: '0.15px',
          lineHeight: 1.28571
        },
        body1Next: {
          letterSpacing: '0.25px',
          lineHeight: 1.25
        },
        body2Next: {
          letterSpacing: '0.15px',
          lineHeight: 1.28571
        },
        button: {
          fontWeight: 700,
          letterSpacing: '1.25px'
        },
        buttonNext: {
          fontWeight: 700,
          letterSpacing: '1.25px'
        },
        caption: {
          letterSpacing: '0.2px',
          lineHeight: 1.33333
        },
        overline: {
          fontWeight: 600,
          fontSize: '0.625rem',
          letterSpacing: '1.5px',
          lineHeight: 1.4
        },
        fontWeightLight: 300,
        fontWeightMedium: 600,
        fontWeightRegular: 400
      },
      palette: {
        primary: {
          main: primaryThemeColor
        },
        secondary: {
          main: secondaryThemeColor,
          contrastText: '#FFF'
        },
        error: {
          main: alertThemeColor
        },
        text: {
          primary: textPrimaryThemeColor,
          secondary: textSecondaryThemeColor
        },
        alert: {
          main: alertThemeColor
        },
        urgent: {
          main: urgentThemeColor
        },
        success: {
          main: successThemeColor
        },
        warning: {
          main: warningThemeColor
        },
        backgroundColor: {
          main: backgroundThemeColor
        },
        mediumGrey: {
          main: mediumGreyThemeColor
        },
        lightGrey: {
          main: lightGreyThemeColor
        },
        crGreen: {
          main: crGreenThemeColor
        },
        darkGrey: {
          main: darkGreyThemeColor
        },
        rpOrange: {
          main: rpOrangeThemeColor
        },
        modernBlack: {
          main: modernBlackThemeColor
        }
      },
      overrides: {
        MuiTypography: {
          root: {
            display: 'block'
          }
        }
      },
      props: {
        MuiButtonBase: {
          TouchRippleProps: {
            style: {
              color: lightGreyThemeColor
            }
          }
        }
      }
    });
  }

  render() {
    return (
      <ThemeProvider theme={this.muiTheme}>
        {this.props.children}
      </ThemeProvider>
    );
  }
}

ThemeWrapper.propTypes = {
  children: PropTypes.node
};

ThemeWrapper.defaultProps = {
  children: null
};

export default ThemeWrapper;
