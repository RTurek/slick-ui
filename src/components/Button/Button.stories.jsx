import React from 'react';
import { boolean, select } from '@storybook/addon-knobs';
import { withActions } from '@storybook/addon-actions';
import { withInfo } from '@storybook/addon-info';
import StarIcon from '@material-ui/icons/Star';
import { BUTTON_SIZES } from '../../styles/sizes';
import { storybookStyles } from '../../../.storybook/storybookTheming';
import Button from './Button';
import StoryWrapper from '../../styles/StoryWrapper';
import buttonNotes from './buttonNotes.md';

const colorOptions = {
  inherit: 'inherit',
  default: 'default',
  primary: 'primary',
  secondary: 'secondary',
  textPrimary: 'textPrimary',
  textSecondary: 'textSecondary',
  alert: 'alert',
  success: 'success',
  warning: 'warning',
  backgroundColor: 'backgroundColor',
  mediumGrey: 'mediumGrey',
  lightGrey: 'lightGrey'
};

const textColor = {
  inherit: 'inherit',
  default: 'default',
  primary: 'primary',
  secondary: 'secondary',
  textPrimary: 'textPrimary',
  textSecondary: 'textSecondary',
  alert: 'alert',
  success: 'success',
  warning: 'warning',
  backgroundColor: 'backgroundColor',
  mediumGrey: 'mediumGrey',
  lightGrey: 'lightGrey',
  contrastPrimary: 'contrastPrimary',
  contrastSecondary: 'contrastSecondary',
  white: 'white',
  black: 'black'
};

export default {
  title: 'Core / Button',

  decorators: [
    withInfo({
      text: buttonNotes,
      header: true,
      inline: true,
      source: false,
      styles: storybookStyles,
      propTables: [Button],
      propTablesExclude: [Button],
      excludedPropTypes: ['theme']
    }),
    withActions('mouseover', 'click')
  ]
};

export const button = () => (
  <StoryWrapper>
    <div style={{ backgroundColor: 'white', padding: '50px', display: 'grid', gridTemplateRows: '1fr', gridRowGap: '20px', justifyItems: 'center' }}>
      <Button
        color={select('Color', colorOptions, 'primary')}
        textColor={select('text Color', textColor, 'white')}
        disabled={boolean('Disabled', false)}
        fullWidth={boolean('fullWidth', true)}
        square={boolean('square', true)}
        makeItPop={boolean('makeItPop', true)}
        size={select('size', BUTTON_SIZES, 'medium')}
        variant="contained"
        startIcon={boolean('startIcon', true) ? <StarIcon /> : null}
        endIcon={boolean('endIcon', false) ? <StarIcon /> : null}
      >
        Contained
      </Button>
      <Button
        color={select('Color', colorOptions, 'primary')}
        textColor={select('text Color', textColor, 'white')}
        disabled={boolean('Disabled', false)}
        fullWidth={boolean('fullWidth', true)}
        square={boolean('square', true)}
        makeItPop={boolean('makeItPop', true)}
        size={select('size', BUTTON_SIZES, 'medium')}
        variant="outlined"
        startIcon={boolean('startIcon', true) ? <StarIcon /> : null}
        endIcon={boolean('endIcon', false) ? <StarIcon /> : null}
      >
        Outlined
      </Button>
      <Button
        color={select('Color', colorOptions, 'primary')}
        textColor={select('text Color', textColor, 'white')}
        disabled={boolean('Disabled', false)}
        fullWidth={boolean('fullWidth', true)}
        square={boolean('square', true)}
        makeItPop={boolean('makeItPop', true)}
        size={select('size', BUTTON_SIZES, 'medium')}
        variant="light"
        startIcon={boolean('startIcon', true) ? <StarIcon /> : null}
        endIcon={boolean('endIcon', false) ? <StarIcon /> : null}
      >
        Light
      </Button>
      <Button
        color={select('Color', colorOptions, 'primary')}
        textColor={select('text Color', textColor, 'white')}
        disabled={boolean('Disabled', false)}
        fullWidth={boolean('fullWidth', true)}
        square={boolean('square', true)}
        makeItPop={boolean('makeItPop', true)}
        size={select('size', BUTTON_SIZES, 'medium')}
        variant="text"
        startIcon={boolean('startIcon', true) ? <StarIcon /> : null}
        endIcon={boolean('endIcon', false) ? <StarIcon /> : null}
      >
        Text
      </Button>
      <Button
        color={select('Color', colorOptions, 'primary')}
        textColor={select('text Color', textColor, 'white')}
        disabled={boolean('Disabled', false)}
        fullWidth={boolean('fullWidth', true)}
        square={boolean('square', true)}
        makeItPop={boolean('makeItPop', true)}
        size={select('size', BUTTON_SIZES, 'medium')}
        variant="contained"
        iconPadding
      >
        <StarIcon />
      </Button>
      <Button
        color={select('Color', colorOptions, 'primary')}
        textColor={select('text Color', textColor, 'white')}
        disabled={boolean('Disabled', false)}
        fullWidth={boolean('fullWidth', true)}
        square={boolean('square', true)}
        makeItPop={boolean('makeItPop', true)}
        size={select('size', BUTTON_SIZES, 'medium')}
        variant="outlined"
        iconPadding
      >
        <StarIcon />
      </Button>
      <Button
        color={select('Color', colorOptions, 'primary')}
        textColor={select('text Color', textColor, 'white')}
        disabled={boolean('Disabled', false)}
        fullWidth={boolean('fullWidth', true)}
        square={boolean('square', true)}
        makeItPop={boolean('makeItPop', true)}
        size={select('size', BUTTON_SIZES, 'medium')}
        variant="light"
        iconPadding
      >
        <StarIcon />
      </Button>
      <Button
        color={select('Color', colorOptions, 'primary')}
        textColor={select('text Color', textColor, 'white')}
        disabled={boolean('Disabled', false)}
        fullWidth={boolean('fullWidth', true)}
        square={boolean('square', true)}
        makeItPop={boolean('makeItPop', true)}
        size={select('size', BUTTON_SIZES, 'medium')}
        variant="text"
        iconPadding
      >
        <StarIcon />
      </Button>
    </div>
  </StoryWrapper>
);

button.story = {
  name: 'Button'
};
