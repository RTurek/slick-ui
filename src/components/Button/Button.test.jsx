import React from 'react';
import renderer from 'react-test-renderer';
import Button from './Button';
import ThemeWrapper from '../../styles/ThemeWrapper';

test('Button should render', () => {
  const component = (
    <ThemeWrapper>
      <Button
        color="primary"
      >
        This Is A Test
      </Button>
    </ThemeWrapper>
  );
  const tree = renderer.create(component).toJSON();
  expect(tree).toMatchSnapshot();
});
