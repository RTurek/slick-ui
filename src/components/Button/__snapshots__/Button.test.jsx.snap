// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render 1`] = `
<button
  className="MuiButtonBase-root MuiButton-root MuiButton-contained Button__StyledButton-sc-1kiwsh8-0 iQlUci react-css"
  disabled={false}
  onBlur={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex={0}
  type="button"
>
  <span
    className="MuiButton-label"
  >
    This Is A Test
  </span>
</button>
`;
